core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java

```diff
@@ -290,7 +290,7 @@ public static void checkActualAndReferenceFiles() {
    */
   @Test void testAsOfCast() {
     final String sql = "SELECT * "
-        + "FROM (SELECT deptno % 10 as m, CAST(deptno AS BIGINT) as deptno FROM dept) D\n"
+        + "FROM (SELECT CAST(deptno % 10 AS BIGINT) as m, CAST(deptno AS BIGINT) as deptno FROM dept) D\n"
```

Comment:
@mihaibudiu If you have time please review this unit test again.
If I don't change it, `D.m = E.empno` will become `(cast(D.m as BIGINT) = E.empno)` so it throws an exception.

Reply:
Maybe this is cause by the recent changes I made to the TypeCoercion which got merged into main: https://github.com/apache/calcite/pull/3998
In this case this change is fine.
The ASOF join seems a bit brittle. I have followed the spec from Snowflake for its definition, but I wonder whether we can improve that.

Reply:
The exception is `ASOF JOIN condition must be a conjunction of equality comparisons`. The ON condition will converted to `CAST(D.M AS BIGINT) = E.EMPNO` if we don't change the D.M type. 
Before this PR, Integer and BigInt don't need to cast. But now it needs.
Before #3998, we don't need to make this change.

Reply:
Adding the cast is fine for now. The spec for ASOF requires it, unfortunately.
Today the validator checks the shape of the ASOF join before optimizations.
But optimizations may change it. Ideally it should only be checked after optimizations.

Reply:
If you don't think we should solve it in this PR, I am preparing to merge it.

Reply:
yes, it is completely unrelated to this PR.